### DISCUSSION TITLE
PHP Version?

### DISCUSSION THREAD

**Discussion by Ruok2bu** [firefly-iii/firefly-iii#5913] [Firefly III]
Which version of PHP works with firefly? I ask because after trying to install with composer with PHP 7.4 i get the following error:

$ composer create-project grumpydictator/firefly-iii --no-dev --prefer-dist firefly-iii 5.6.16
Creating a "grumpydictator/firefly-iii" project at "./firefly-iii"
Info from https://repo.packagist.org: #StandWithUkraine

                                                                               
  [InvalidArgumentException]                                                   
  Could not find package grumpydictator/firefly-iii with version 5.6.16 in a   
  version installable using your PHP version, PHP extensions and Composer ver  
  sion.                                                                        
                                                                               

create-project [-s|--stability STABILITY] [--prefer-source] [--prefer-dist] [--prefer-install PREFER-INSTALL] [--repository REPOSITORY] [--repository-url REPOSITORY-URL] [--add-repository] [--dev] [--no-dev] [--no-custom-installers] [--no-scripts] [--no-progress] [--no-secure-http] [--keep-vcs] [--remove-vcs] [--no-install] [--ignore-platform-req IGNORE-PLATFORM-REQ] [--ignore-platform-reqs] [--ask] [--] [<package>] [<directory>] [<version>]

**Comment by Lawri-van-Buel** [comment]
Based on https://packagist.org/packages/grumpydictator/firefly-iii you need php >=8 can you try php 8.0 or php 8.1?

**Comment by Ruok2bu** [comment]
I found fix in: https://github.com/firefly-iii/firefly-iii/issues/1651

**Reply by JC5** [reply]
Run it with ` --verbose`, you may get some more info. You're probably missing PHP packages.

**Comment by Ruok2bu** [comment]
Now it shows this error:

$ composer create-project grumpydictator/firefly-iii --no-dev --verbose
Creating a "grumpydictator/firefly-iii" project at "./firefly-iii"
Installing grumpydictator/firefly-iii (4.0.1)
  - Installing grumpydictator/firefly-iii (4.0.1): Extracting archive
Created project in /**************/**************/**************/budget/firefly-iii
> post-root-package-install: php -r "file_exists('.env') || copy('.env.example', '.env');"
Installing dependencies from lock file
Verifying lock file contents can be installed on current platform.
Dependency resolution completed in 0.000 seconds
Your lock file does not contain a compatible set of packages. Please run composer update.
a
  Problem 1
    - doctrine/cache is locked to version v1.6.0 and an update of this package was not requested.
    - doctrine/cache v1.6.0 requires php ~5.5|~7.0 -> your php version (8.1.3) does not satisfy that requirement.
  Problem 2
    - doctrine/common is locked to version v2.6.1 and an update of this package was not requested.
    - doctrine/common v2.6.1 requires php ~5.5|~7.0 -> your php version (8.1.3) does not satisfy that requirement.
  Problem 3
    - doctrine/common v2.6.1 requires php ~5.5|~7.0 -> your php version (8.1.3) does not satisfy that requirement.
    - doctrine/dbal v2.5.5 requires doctrine/common >=2.4,<2.7-dev -> satisfiable by doctrine/common[v2.6.1].
    - doctrine/dbal is locked to version v2.5.5 and an update of this package was not requested.

**Reply by JC5** [reply]
Yes but now you're trying to install a very old version.

**Comment by Ruok2bu** [comment]
When i followed the instructions in https://github.com/firefly-iii/firefly-iii/issues/1651 where it says not to specify the version, installation get's farther. I understand that the errors above are because its installing an older version of Firefly which isnt compatible with PHP 8.0

Thing is, i dont know how to get it to install the current version because it gives errors.  I have multiple versions of PHP installed on my web server, PHP 5.6 is also installed. Thing is i cant uninstall it because then PHP 8.1 is also uninstalled (i use Ondrej's PHP repo and for whatever reason its set up like that - he has bleeding edge updates for PHP in Debian OS's).

**Reply by JC5** [reply]
The ticket you refer to is over 3 years old. Many things have changed in the mean time. 

As per the documentation, Firefly III requires php 8.0 (at the time of writing). So, use 8.0 and use the command mentioned here:

https://docs.firefly-iii.org/firefly-iii/installation/self_hosted/

There's not much more I can do for you.

**Comment by Ruok2bu** [comment]
I downgraded to 8.0, and removed 8.1, Installed all the required extensions. Php is working elsewhere on my site. But when i follow those instructions i get this error (i ran it as root since the documentation said to if i get errors):

$ sudo composer create-project grumpydictator/firefly-iii --no-dev --prefer-dist firefly-iii 5.6.16 --verbose
Do not run Composer as root/super user! See https://getcomposer.org/root for details
Continue as root/super user [yes]? yes
Creating a "grumpydictator/firefly-iii" project at "./firefly-iii"

                                                                               
  [InvalidArgumentException]                                                   
  Could not find package grumpydictator/firefly-iii with version 5.6.16 in a   
  version installable using your PHP version, PHP extensions and Composer ver  
  sion.                                                                        
                                                                               

Exception trace:
 () at phar:///usr/local/bin/composer/src/Composer/Command/CreateProjectCommand.php:436
 Composer\Command\CreateProjectCommand->installRootPackage() at phar:///usr/local/bin/composer/src/Composer/Command/CreateProjectCommand.php:207
 Composer\Command\CreateProjectCommand->installProject() at phar:///usr/local/bin/composer/src/Composer/Command/CreateProjectCommand.php:168
 Composer\Command\CreateProjectCommand->execute() at phar:///usr/local/bin/composer/vendor/symfony/console/Command/Command.php:245
 Symfony\Component\Console\Command\Command->run() at phar:///usr/local/bin/composer/vendor/symfony/console/Application.php:835
 Symfony\Component\Console\Application->doRunCommand() at phar:///usr/local/bin/composer/vendor/symfony/console/Application.php:185
 Symfony\Component\Console\Application->doRun() at phar:///usr/local/bin/composer/src/Composer/Console/Application.php:336
 Composer\Console\Application->doRun() at phar:///usr/local/bin/composer/vendor/symfony/console/Application.php:117
 Symfony\Component\Console\Application->run() at phar:///usr/local/bin/composer/src/Composer/Console/Application.php:131
 Composer\Console\Application->run() at phar:///usr/local/bin/composer/bin/composer:83
 require() at /usr/local/bin/composer:29

create-project [-s|--stability STABILITY] [--prefer-source] [--prefer-dist] [--prefer-install PREFER-INSTALL] [--repository REPOSITORY] [--repository-url REPOSITORY-URL] [--add-repository] [--dev] [--no-dev] [--no-custom-installers] [--no-scripts] [--no-progress] [--no-secure-http] [--keep-vcs] [--remove-vcs] [--no-install] [--ignore-platform-req IGNORE-PLATFORM-REQ] [--ignore-platform-reqs] [--ask] [--] [<package>] [<directory>] [<version>]

**Reply by JC5** [reply]
That's weird. What's the output of the following commands?

```
sudo php -v
php -v
```

**Comment by Ruok2bu** [comment]
PHP 8.0.16 (cli) (built: Feb 21 2022 14:47:25) ( NTS )
Copyright (c) The PHP Group
Zend Engine v4.0.16, Copyright (c) Zend Technologies
    with Zend OPcache v8.0.16, Copyright (c), by Zend Technologies

PHP 8.0.16 (cli) (built: Feb 21 2022 14:47:25) ( NTS )
Copyright (c) The PHP Group
Zend Engine v4.0.16, Copyright (c) Zend Technologies
    with Zend OPcache v8.0.16, Copyright (c), by Zend Technologies

**Comment by Ruok2bu** [comment]
Here's some info about the OS if that's helpful:

OS: Debian 10
PHP Repo: Ondřej Surý https://launchpad.net/~ondrej/+archive/ubuntu/php
Web Server: Abyss Web Server  (it combines elements of IIS and apache)

**Reply by JC5** [reply]
Final thing you can try is to add `--ignore-platform-reqs`, which may make it run regardless. 

I'm curious why you're using such an obscure web server though.

**Reply by kitto-someday92** [reply]
Hello, 

I've also tried to install firefly-iii on a Linux 22.04 machine but i received the the same error as Ruok2bu. I did manage to avoid it by using the parameter you mentioned in this post but unfortunately when I try to run php artisan migrate:refresh --seed i get the following error:

```
php artisan migrate:refresh --seed
PHP Fatal error:  Uncaught UnexpectedValueException: The stream or file "/var/www/firefly-iii/storage/logs/laravel.log" could not be opened in append mode: Failed to open stream: Permission denied
The exception occurred while attempting to log: Unable to create configured logger. Using emergency logger.
Context: {"exception":{}} in /var/www/firefly-iii/vendor/monolog/monolog/src/Monolog/Handler/StreamHandler.php:146
Stack trace:
#0 /var/www/firefly-iii/vendor/monolog/monolog/src/Monolog/Handler/AbstractProcessingHandler.php(48): Monolog\Handler\StreamHandler->write()
#1 /var/www/firefly-iii/vendor/monolog/monolog/src/Monolog/Logger.php(342): Monolog\Handler\AbstractProcessingHandler->handle()
#2 /var/www/firefly-iii/vendor/monolog/monolog/src/Monolog/Logger.php(622): Monolog\Logger->addRecord()
#3 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Log/Logger.php(183): Monolog\Logger->emergency()
#4 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Log/Logger.php(58): Illuminate\Log\Logger->writeLog()
#5 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Log/LogManager.php(138): Illuminate\Log\Logger->emergency()
#6 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Support/helpers.php(302): Illuminate\Log\LogManager->Illuminate\Log\{closure}()
#7 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Log/LogManager.php(140): tap()
#8 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Log/LogManager.php(119): Illuminate\Log\LogManager->get()
#9 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Log/LogManager.php(702): Illuminate\Log\LogManager->driver()
#10 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Foundation/Exceptions/Handler.php(274): Illuminate\Log\LogManager->log()
#11 /var/www/firefly-iii/app/Exceptions/Handler.php(144): Illuminate\Foundation\Exceptions\Handler->report()
#12 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Foundation/Console/Kernel.php(367): FireflyIII\Exceptions\Handler->report()
#13 /var/www/firefly-iii/vendor/laravel/framework/src/Illuminate/Foundation/Console/Kernel.php(131): Illuminate\Foundation\Console\Kernel->reportException()
#14 /var/www/firefly-iii/artisan(37): Illuminate\Foundation\Console\Kernel->handle()
#15 {main}
  thrown in /var/www/firefly-iii/vendor/monolog/monolog/src/Monolog/Handler/StreamHandler.php on line 146
```

The results of php -v and and sudo php -v are below

php -v:

`PHP 8.0.19 (cli) (built: May 17 2022 18:49:30) ( NTS )
Copyright (c) The PHP Group
Zend Engine v4.0.19, Copyright (c) Zend Technologies
    with Zend OPcache v8.0.19, Copyright (c), by Zend Technologies
`

sudo php -v:

`PHP 8.0.19 (cli) (built: May 17 2022 18:49:30) ( NTS )
Copyright (c) The PHP Group
Zend Engine v4.0.19, Copyright (c) Zend Technologies
    with Zend OPcache v8.0.19, Copyright (c), by Zend Technologies
`

If it matters I'm running a LAMP stack and the mysql DB runs on a different machine. Could you help me by giving me a hint on what should i do?

**Comment by Ruok2bu** [comment]
Because i regularly add/remove small sites and editing nginx config files is a pain in the ass.

Also it finally installed! Trying to configure it now.  Many thanks!

**Comment by martinkirch** [comment]
If someone encounters this problem too : adding the `--ignore-platform-reqs` seems the way to go. Might be worth mentioning in the `self_hosted` section of the documentation.

**Reply by JC5** [reply]
I'd rather not, to be honest. You could run into any weird issue.
